### ISSUE TITLE
Facilitate the use of custom Datasets

### ISSUE THREAD

**Issue by BenjaminBossan** [dnouri/skorch#422]
Custom `Dataset`s are somewhat second class citizens in skorch land. For instance, see this blogpost:

https://nodata.science/no-fit-comparing-high-level-learning-interfaces-for-pytorch.html

I could imagine that others also face problems when they want to transition their existing pytorch code to skorch.

The solution is not straightforward, since we probably cannot change skorch in a way that allows to use any kind of `Dataset`. However, we can probably give more helpful error messages. E.g., when `__getitem__` does not return exactly two values, we could point that out to the user and tell them how to fix the problem.

I personally work mostly with numpy arrays or dicts of arrays, thus I rarely encounter this sort of problem. So if anyone can share their trouble they had with `Dataset`s, that would be helpful.

**Comment by thomasjpfan** [comment]
I exclusively use `Dataset` and was able to avoid issues by getting familiar with the skorch codebase. One way to work around the two value problem, is to gather multiple x values as a tuple. Doing it this way, would require all other code to be able to handle tuples.

**Comment by BenjaminBossan** [comment]
> One way to work around the two value problem, is to gather multiple x values as a tuple.

Is your suggestion to gather all outputs from `__getitem__` and if there are more than 2, put all but the last in a tuple as `X`?

> Doing it this way, would require all other code to be able to handle tuples.

Implicitly, we support tuples (e.g. in `to_tensor`) though not officially. Did you encounter problems with tuples?

**Comment by thomasjpfan** [comment]
> Is your suggestion to gather all outputs from __getitem__ and if there are more than 2, put all but the last in a tuple as X?

Yes this is what I meant. 

> Did you encounter problems with tuples?

Nope. As a concrete example, this works:

```python
from torch.utils.data import DataLoader, Dataset
from skorch import NeuralNetRegressor
import torch.nn as nn
import torch

class MyDS(Dataset):
    def __init__(self):
        self.X = [(torch.randn(10), torch.randn(10)) for _ in range(100)]
        self.Y = torch.randn(100)
    def __getitem__(self, idx):
        return self.X[idx], self.Y[idx]
    def __len__(self):
        return len(self.X)

class MyModule1(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear0 = nn.Linear(10, 1)
        self.linear1 = nn.Linear(10, 1)
    def forward(self, X):
        return self.linear0(X[0]) + self.linear1(X[1])

dataset = MyDS()
net = NeuralNetRegressor(MyModule1)
_ = net.fit(dataset, y=None)
``` 

For reference, dictionaries also work (although kind of hidden):

```python
class MyDS_Dict(Dataset):
    def __init__(self):
        self.X = [{"X1": torch.randn(10), "X2": torch.randn(10)} for _ in range (100)]
        self.Y = torch.randn(100)
    def __getitem__(self, idx):
        return self.X[idx], self.Y[idx]
    def __len__(self):
        return len(self.X)

class MyModule2(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear0 = nn.Linear(10, 1)
        self.linear1 = nn.Linear(10, 1)
    def forward(self, X1, X2):
        return self.linear0(X1) + self.linear1(X2)

dataset_dict = MyDS_Dict()
net = NeuralNetRegressor(MyModule2)
net.fit(dataset_dict, y=None)
```

Notice how `MyModule2` accepts variables `X1` and `X2`, which are the keys to the dataset's `self.X` list of dictionaries.

**Comment by BenjaminBossan** [comment]
So there are a couple of ways forward here. Each time we iterate over `self.get_iterator(dataset, ...)`, we count the number of return values, then:

1. If there are exactly 2, fine. If more than 2, group all but the last in a tuple. If less than 2, idk. Then we would hope that the user properly handles the grouping, as in the first example above.
2. If there are exactly 2, fine. If not, raise an error with an explanation and a link to the docs.

The advantage with 1 is that it could just work. But I'm not sure if many people would define their `forward` correctly, and if they don't, it could lead to frustration*. Therefore, I have a slight preference for 2. What do others think?

*We could change the `infer` call to be `self.module_(*X, **fit_params)` if `X` is a tuple. I imagine that this would fit the user implementation more often.

**Comment by thomasjpfan** [comment]
I prefer option 2. 

> We could change the infer call to be self.module_(*X, **fit_params) if X is a tuple.

This may cause some confusion. I think letting the user unpack the tuples themselves would be more explicit. A user would need to do unpacking for all callbacks that are passed X, such as in the `on_train_begin`.

**Comment by thomasjpfan** [comment]
In reference to the OP's link, our [docs](https://skorch.readthedocs.io/en/latest/user/dataset.html?highlight=%22PyTorch%20DataLoaders%20by%20%22#id1) discourage the use of PyTorch's dataset object. Since skorch works with the PyTorch dataset (with some restrictions), that section should include a link to how to use a PyTorch dataset. The link could be a new item on the FAQ.

**Comment by Venka97** [comment]
Hey so if my __getitem__  returns a dict of  {'image1':img1, 'image2':img2, 'mask':mask} , how can I use it with skorch?

**Comment by BenjaminBossan** [comment]
@vThe only condition is that you return two values, X and y. If you don't have a y, just return a dummy value like here:

https://github.com/dnouri/skorch/blob/a5ad39a28527c66c4c0fcfb73d348b976657fba7/skorch/dataset.py#L202

Note that in case of a dict, the keys must match the argument names of the `forward` method of your module, as described in the [docs](https://skorch.readthedocs.io/en/latest/user/dataset.html#id1).
